stan/math/prim/fun/log_mix.hpp

```diff
@@ -157,38 +91,19 @@ return_type_t<T_theta, T_lam> log_mix(const T_theta& theta,
   check_finite(function, "theta", theta);
   check_consistent_sizes(function, "theta", theta, "lambda", lambda);
 
-  scalar_seq_view<T_theta> theta_vec(theta);
-  T_partials_vec theta_dbl(N);
-  for (int n = 0; n < N; ++n) {
-    theta_dbl[n] = value_of(theta_vec[n]);
-  }
-
-  scalar_seq_view<T_lam> lam_vec(lambda);
-  T_partials_vec lam_dbl(N);
-  for (int n = 0; n < N; ++n) {
-    lam_dbl[n] = value_of(lam_vec[n]);
-  }
+  const auto& theta_dbl = value_of(as_column_vector_or_scalar(theta));
+  const auto& lam_dbl = value_of(as_column_vector_or_scalar(lambda));
 
   T_partials_return logp = log_sum_exp((log(theta_dbl) + lam_dbl).eval());
```

Comment:
The ```.eval()``` shouldn't be necessary anymore

Reply:
If I remove that it produces wrong result. After some digging I found the problem. Rev implementation of `log_sum_exp` doesn not work correctly with expressions. 

It allocates the vari for the result before evaluating the expression. Evaluating the expression creates some varis for intermediate results. I think this results in varis being on the chain_stack in wrong order.

I could evaluate (or use an `Eigen::Ref`) in the the rev `log_sum_exp`. This works, but makes an extra copy. Also I think we will find similar issues in other rev functions. I will need to think a bit about this.